lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -702,20 +710,95 @@ def _get_run_job_params_from_existing_job(job: Job, user_id: str) -> Dict:
         # Then the next fields are job inputs top level requirements, app run parameters, and scheduler resource requirements
         return run_job_params
 
+    def _check_ws_perms(
+        self, runjob_params, new_batch_job, batch_params, as_admin=False
+    ):
+        """
+        Check a single job, a single batch job, or a retry_multiple request with a mix of different jobs.
+        """
+        if as_admin:
+            return self.sdkmr.check_as_admin(requested_perm=JobPermissions.WRITE)
+        # Batch Param runs
+        if new_batch_job:
+            if batch_params:
+                return self._check_workspace_permissions(batch_params.get("wsid"))
+        # Single job runs
+        elif isinstance(runjob_params, dict):
+            return self._check_workspace_permissions(runjob_params.get("wsid"))
+        # Multiple independent job runs, think retry_multiple()
+        elif isinstance(runjob_params, list):
+            return self._check_workspace_permissions_list(
+                [job_param.get("wsid") for job_param in runjob_params]
+            )
+        else:
+            raise IncorrectParamsException(
```

Comment:
Added `        if type(params) != dict:
            raise IncorrectParamsException("params must be a mapping")` 
to run()

So if the only way to get to this function is through 

run->preflight->_check_ws_perms
or
run_batch->Preflight->_check_ws_perms

And both of them check to see that the inputs are a dict or a list first, how do we test this exception by only testing the public api?

Reply:
If code can't be reached via the public API then the code isn't needed and can be deleted

Reply:
Hmmmmmmm something about this just seems funny but alright, will just delete it then.

Reply:
It could be seen as future proofing the code for new uses of the method. If that's the intent I would drop the final else and just end the method with
```
assert 0, "called method with invalid params"
```
 or something like that

Reply:
Is that better than throwing an exception? And the code is still not covered in that case as well, right?

Reply:
yeah, because you can have the interpreter remove asserts.

Reply:
Hmm, now I'm reviewing this again this definitely seems like a stumbling block for future coders - if they submit an uncovered scenario it'll just drop through the method and seem like the perms were checked when they weren't. I'd at least put the assert in to protect against that.

Reply:
I've never used assert outside of tests because they can get removed during compilation.  So again I don't understand why you want to assert here instead of throwing an exception

Reply:
So if you want to use an exception that's fine, it's not worth the book we've written at this point :P

That being said, this is exactly what asserts are for - when you're debugging / testing code, you leave the asserts in to catch mistakes. When you're ready for deploy, you can compile them out.